### PR TITLE
EE-1100: Transfer error handling

### PR DESCRIPTION
**Pull request by mpapierski**
Ref: https://casperlabs.atlassian.net/browse/EE-1100

- Propagates mint's ApiError to the caller through ext_ffi
- Replaces `Result<(), ()>` with proper error handling
- Augments Auction errors with different errors whenever transfer fails so it's not always `Transfer` but `Auction(something happened during transfer)`
- Adds assemblyscript support for said feature

Now `ApiError::Transfer` should be considered only for really unhandled transfer errors such as failed to call the contract, can't reconstruct `TransferredTo` etc. which should be rare cause.

### CODE REVIEW

**Review by Fraser999 — APPROVED** [review]
This looks good to me (bar a single nitpick).  However, I didn't thoroughly review the typescript changes, so it'd probably be worth waiting for a second reviewer before merging this.

File: smart_contracts/contract/src/contract_api/system.rs
Comment (EdHastingsCasperLabs) [inline smart_contracts/contract/src/contract_api/system.rs:124]: nitpick: // Propagate error (if any)

File: types/src/system_contract_errors/auction.rs
Comment (EdHastingsCasperLabs) [inline types/src/system_contract_errors/auction.rs:126]: Why do we need this variant?

**Review by EdHastingsCasperLabs — APPROVED** [review]
see nitpicks

File: types/src/system_contract_errors/auction.rs
Comment (mpapierski) [reply types/src/system_contract_errors/auction.rs:126]: Before this change

```
#[cfg(test)]
const MAX_ERROR_VALUE
```

was hardcoded and `enum Error` values grew, but tests didnt reflect that. With this `MAX_ERROR_VALUE` would be always highest value and thus all values will be tested (if we'd extend error enum in future)

**Pull request merged**
